File: src/main/java/year2019/Day01.java
package year2019;

import aoc.IAocTask;

import java.util.List;

public class Day01 implements IAocTask {
    @Override
    public String getFileName() {
        return "aoc2019/input_01.txt";
    }

    @Override
    public void solvePartOne(List<String> lines) {
        Integer sum = lines.stream()
                .map(Integer::valueOf)
                .map(mass -> mass / 3 - 2)
                .reduce(Integer::sum)
                .orElse(-1);
        System.out.println(sum);
    }

    @Override
    public void solvePartTwo(List<String> lines) {
        Integer sum = lines.stream()
                .map(Integer::valueOf)
                .map(this::getRecursiveSum)
                .reduce(Integer::sum)
                .orElse(-1);
        System.out.println(sum);

    }

    private int getRecursiveSum(Integer mass) {
        int sum = 0;
        mass = mass / 3 - 2;
        while (mass > 0) {
            sum += mass;
            mass = mass / 3 - 2;
        }
        return sum;
    }
}

File: src/main/kotlin/year2022/Day01.kt
package year2022

import aoc.IAocTaskKt

class Day01 : IAocTaskKt{
    override fun getFileName(): String = "aoc2022/input_01.txt"

    override fun solvePartOne(lines: List<String>) {
        val caloriesPerElf = caloriesPerElf(lines)
        val maxCalories = caloriesPerElf.maxOf { it }
        println(maxCalories)
    }

    private fun caloriesPerElf(lines: List<String>) =
        lines.fold(mutableListOf(mutableListOf<Int>())) { elvesCalories, line ->
            if (line.isNotEmpty()) {
                elvesCalories.last().add(line.toInt())
            } else {
                elvesCalories.add(mutableListOf())
            }
            elvesCalories
        }.map { it.sum() }

    override fun solvePartTwo(lines: List<String>) {
        println(caloriesPerElf(lines).sortedDescending().subList(0, 3)
            .sumOf { it })
    }
}
